New Block: Add Read Progress block

Infinite-Null opened the pull request:
Related PR: https://github.com/WordPress/gutenberg/pull/68652
PR Comment: https://github.com/WordPress/gutenberg/pull/68652#issuecomment-2654482297 

## What?
This PR introduces a new Read Progress Bar block that visually indicates a reader’s progress through the content as they scroll down the page.

## Why?
- Enhances user experience by providing a clear visual cue for reading progress.
- Useful for long-form content, helping readers track their position.
- Aligns with modern web design practices seen in blogs and news websites.

## Screenshots or screencast 

https://github.com/user-attachments/assets/88035d9a-9e44-4ad2-92f1-d87b22739921

Infinite-Null committed:
Add Read Progress block

- Create new core/read-progress block for tracking reading progress
- Add styles with scroll animation timeline for tracking progress
- Register block in PHP and include in block registry
- Implement edit and save components with fixed position styling
- Add block.json with metadata and configuration
- Include proper admin bar positioning for WordPress admin interface
- Create animation keyframes for smooth progress visualization
- Add block to core block list in both PHP and JS registrations

Infinite-Null committed:
Add color customization to Read Progress block

- Add backgroundColor and progressColor attributes to block.json with defaults
- Add color settings panel in the block inspector controls
- Update edit.js to use color attributes and handle color changes
- Update save.js to apply selected colors to the rendered block
- Ensure color settings persist through block

Infinite-Null committed:
Add read-progress block scroll-timeline polyfill for it

- Added interactivity support to the read-progress block by enabling the `interactivity` property in `block.json`.
- Ensured the progress bar is animated correctly using the `ScrollTimeline` polyfill.
- Added a new `view.js` file to handle the scroll-based animation logic for the progress bar.
- Included a polyfill utility to ensure compatibility with browsers that do not natively support `ScrollTimeline`.

Infinite-Null committed:
Fix Coding standard

Infinite-Null committed:
Add configurable height option to read progress block

- Added height property to block.json with default value of 15 pixels
- Implemented height control in the editor UI with a range slider
- Updated both edit.js and save.js to dynamically set the progress bar height
- Added Tools Panel with reset functionality for height settings

Infinite-Null committed:
Add reading progress icon and update its usage in the block

- Added a new read-progress icon SVG in the icons library
- Updated the read-progress block to use the dedicated icon instead of reusing the link icon
- Designed the icon to visually represent a read progress bar

Infinite-Null committed:
Add block selector and improve description

- Updated block description for improved clarity
- Added a block selector for the 'Read Progress' block in the editor
- Styled the selector with a centered gray background for better visibility
- Ensured the selector text is translatable for i18n support
- Improved code readability by consistent formatting

Infinite-Null committed:
Merge branch 'trunk' into add-block/read-progress-block

Infinite-Null committed:
Read Progress: Added position control to toggle between top and bottom

- Added position attribute to block.json with 'top' as default
- Implemented position control using CustomSelectControl in edit.js
- Updated save.js to handle position styling for bottom placement
- Added position reset functionality in Tools Panel

Infinite-Null committed:
Remove read progress icon to fix tests

Infinite-Null committed:
Add isInverted option to read progress block

- Added isInverted boolean attribute with default false
- Added ToggleControl in editor UI for configuring inverted direction
- Updated transformOrigin property based on isInverted value
- Removed hardcoded transform-origin from style.scss

Infinite-Null committed:
Conditionally load polyfill to optimize performance

adamsilverstein commented on docs/reference-guides/core-blocks.md:
```diff
@@ -805,6 +805,15 @@ Displays the link of a post, page, or any other content-type. ([Source](https://
 
+## Read Progress
+
+Displays a read progress. Useful for tracking read progress. ([Source](https://github.com/WordPress/gutenberg/tree/trunk/packages/block-library/src/read-progress))
```
Maybe "Displays a reading progress bar"?

Infinite-Null replied:
Updated the name to `reading progress bar` throughout

adamsilverstein commented on packages/block-library/src/index.js:
```diff
@@ -21,37 +21,39 @@ import {
-import * as avatar from './avatar';
 import * as audio from './audio';
+import * as avatar from './avatar';
+import * as reusableBlock from './block';
```
Some of the changes in the this file seem unintentional, can you revert those changes please and only include the new import. It looks like these belong in alphabetical order.

Infinite-Null replied:
Updated the file to have only the necessary changes.

adamsilverstein commented on packages/block-library/src/index.js:
```diff
@@ -101,8 +103,8 @@ import * as queryPaginationPrevious from './query-pagination-previous';
 import * as quote from './quote';
-import * as reusableBlock from './block';
 import * as readMore from './read-more';
+import * as readProgress from './read-progress';
```
this should be the only change in this file

adamsilverstein commented on packages/block-library/src/read-progress/block.json:
```diff
@@ -0,0 +1,36 @@
+	"name": "core/read-progress",
+	"category": "theme",
+	"title": "Read Progress",
+	"description": "Displays a read progress. Useful for tracking read progress.",
```
"Displays a read progress" isn't quite right, "read" is past tense so I would say instead it displays a "reading progress bar" or a "reading progress indicator".

adamsilverstein commented on packages/block-library/src/read-progress/block.json:
```diff
@@ -0,0 +1,36 @@
+		},
+		"progressColor": {
+			"type": "string",
+			"default": "#1E1E1E"
```
Curious how you chose these colors? Can the progress bar inherit a theme color?

Infinite-Null replied:
I tried to find a way to pass theme's primary color, but couldn’t find a way to achieve it. This color is default $gray-900 color.

https://github.com/WordPress/gutenberg/blob/efa776333c9ad2a9f08a01f0f4f312c1ef5b152a/packages/base-styles/_colors.scss#L9

adamsilverstein replied:
> I tried to find a way to pass theme's primary color, but couldn’t find a way to achieve it.

Interesting, I feel like this would be a great feature if it isn't already available - ie. the ability to have UI components inherit base theme color.

adamsilverstein commented on packages/block-library/src/read-progress/edit.js:
```diff
@@ -0,0 +1,169 @@
+		<div { ...blockProps }>
+			<InspectorControls>
+				<ToolsPanel
+					label={ __( 'Progress Bar Settings' ) }
```
consistent naming "Reading Progress Bar"

Infinite-Null replied:
Updated the name to `reading progress bar` throughout

adamsilverstein commented on packages/block-library/src/read-progress/edit.js:
```diff
@@ -0,0 +1,169 @@
+						label={ __( 'Bar height' ) }
+						isShownByDefault
+						hasValue={ () => height !== 11 }
+						onDeselect={ () => setAttributes( { height: 11 } ) }
```
why does this set height to 11 on deselect?

Infinite-Null replied:
Updated it to `15` as `15` is the default value.

adamsilverstein replied:
Since it isn't clear where this value comes from other reviewers may have this same question. Two ways you could help - either add an inline comment (`/* This is the default value */`) or probably better, move the value into a constant (`const defaultHeight = 15`).

Infinite-Null replied:
Hi @adamsilverstein,
I have updated the PR and created default constants.

```js
/**
 * Default Constants
 */
const DEFAULT_HEIGHT = 15;
const DEFAULT_POSITION = 'top';
```

adamsilverstein commented on packages/block-library/src/read-progress/edit.js:
```diff
@@ -0,0 +1,169 @@
+						isShownByDefault
+						hasValue={ () => position !== 'top' }
+						onDeselect={ () =>
+							setAttributes( { position: 'top' } )
```
why is this setting to top on deselect? in my testing, setting to bottom did not stick - the editor always shows top as selected (although the bar moved to the bottom)

Infinite-Null replied:
The bug was there because I was passing wrong value to `value` prop of `CustomSelectControl`. Now it is fixed!

adamsilverstein commented on packages/block-library/src/read-progress/index.js:
```diff
@@ -0,0 +1,30 @@
+	save,
+	example: {
+		attributes: {
+			label: _x( 'Read Progress', 'block example' ),
```
use same name everywhere

Infinite-Null replied:
Updated the name to `reading progress bar` throughout

adamsilverstein replied:
Great, thanks!

adamsilverstein commented on packages/block-library/src/read-progress/index.php:
```diff
@@ -0,0 +1,38 @@
+ *
+ * @return string Returns the block content.
+ */
+function render_block_read_progress( $attributes, $content ) {
```
fix phpcs error

Infinite-Null replied:
Fixed PHPCS issue

adamsilverstein commented on packages/block-library/src/read-progress/style.scss:
```diff
@@ -0,0 +1,50 @@
+		width: 100%;
+		z-index: 1000;
+
+		.admin-bar & {
```
great, good job catching admin bar spacing

Infinite-Null replied:
Thank You 😁

adamsilverstein commented on packages/block-library/src/read-progress/view.js:
```diff
@@ -0,0 +1,28 @@
+if ( ! ( 'ScrollTimeline' in window ) ) {
```
Nice! this should ensure the script only loads when needed.

Infinite-Null replied:
Thank You 😁

adamsilverstein commented on packages/block-library/src/read-progress/view.js:
```diff
@@ -0,0 +1,28 @@
+	require( './utils/polyfill' );
+}
+
+document.addEventListener( 'DOMContentLoaded', () => {
```
shouldn't this event listener be in the conditional block above? is the event listener needed for the CSS implementation?

Infinite-Null replied:
I added event listener inside the conditional block. The event listener is essential, as it ensures we only manipulate DOM elements, after the HTML is fully parsed.

adamsilverstein reviewed (commented):
This works well in my testing, nice work! Left some specific feedback points, one big one is to settle on a single name to call this feature across UI and docs. I also noticed one UI glitch with the position selector.

Infinite-Null committed:
Rename block and improve settings handling
- Renamed block from `read progress` to `reading progress bar`
- Updated the grammar in the description
- Updated `block-library/src/index.js` to have only necessary changes
- Updated height on deselect from `11px` to `15px`
- Fixed the bug where the value was not updating on `CustomSelectControl`

Infinite-Null committed:
Fix PHPCS issue and refactor `DOMContentLoaded` event listener

Infinite-Null committed:
Merge branch 'trunk' into add-block/read-progress-block

Infinite-Null committed:
Add integration fixtures for the read progress block

Infinite-Null commented:
Hi @adamsilverstein,
Thank you for the feedback's. I’ve addressed all the requested changes. Could you please review the updated PR at your convenience.

Infinite-Null committed:
Refactor default constants for height and position

Infinite-Null committed:
Merge branch 'trunk' into add-block/read-progress-block

adamsilverstein committed:
Merge branch 'trunk' into add-block/read-progress-block

adamsilverstein reviewed (approved):
Looks good!

adamsilverstein commented:
This block might be a good candidate for a canonical block - that is, a block that is maintained by the core team (but not shipped bundled with WordPress or Gutenberg). Ideally A canonical block would be surfaced ahead of third party blocks (eg. from the block inserter, typing scroll... this block would show above other options) - perhaps adding some visual indicator that this is a canonical block. 

Has the idea of canonical blocks - and exactly what they are - been discussed already in an issue or post? If not, I am interested in working on a formal proposal..

cc: @desrosj, @Mamaduka
